lightly/loss/hypersphere_loss.py

```diff
@@ -0,0 +1,96 @@
+"""
+https://arxiv.org/pdf/2005.10242.pdf
+
+NOTE: hypersphere is perhaps bad naming as I am not sure it is the essence;
+alignment-and-uniformity loss perhaps? Does not sound as nice.
+"""
+
+import torch
+
+
+def normalize_l2(x, dim=-1):
+    """Apply l2 normalization along a given axis of a tensor
+
+    Args:
+        x : torch.Tensor, [...], float
+        dim : int
+
+    Returns:
+        torch.Tensor, [...], float
+            tensor l2-normalized along the given axis
+
+    """
+
+    return x / x.norm(p=2, dim=dim, keepdim=True)
+
+
+class HypersphereLoss(torch.nn.Module):
+    """
+
+    Implementation of the loss described in 'Understanding Contrastive Representation Learning through
+    Alignment and Uniformity on the Hypersphere.' [0]
+    [0] Tongzhou Wang. et.al, 2020, ... https://arxiv.org/pdf/2005.10242.pdf
+
+        Examples:
+
+        >>> # initialize loss function
+        >>> loss_fn = HypersphereLoss()
+        >>>
+        >>> # generate two random transforms of images
+        >>> t0 = transforms(images)
+        >>> t1 = transforms(images)
+        >>>
+        >>> # feed through SimSiam model
+        >>> out0, out1 = model(t0, t1)
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(out0, out1)
+
+    """
+
+    def __init__(self, t=1., lam=1., alpha=2.):
+        """Parameters as described in [0]
+
+        Args:
+            t : float
+                Temperature parameter;
+                proportional to the inverse variance of the Gaussians used to measure uniformity
+            lam : float:
+                Weight balancing the alignment and uniformity loss terms
+            alpha : float
+                Power applied to the alignment term of the loss. At its default value of 2,
+                distances between positive samples are penalized in an l-2 sense.
+
+        """
+        super(HypersphereLoss, self).__init__()
+        self.t = t
+        self.lam = lam
+        self.alpha = alpha
+
+    def forward(self, z_a: torch.Tensor, z_b: torch.Tensor) -> torch.Tensor:
+        """
+
+        Args:
+            x : torch.Tensor, [b, d], float
+            y : torch.Tensor, [b, d], float
+
+        Returns:
+            torch.Tensor, [], float
+                scalar loss value
+
+        """
+        # FIXME: this is necessary for the loss to function as advertized,
+        #  but not technically part of the loss i feel
+        #  as it needs to be applied during inference as well
+        #  leave that responsibility to the end user?
```

Comment:
I would leave it here since it's needed for the loss calculation. For inference, I would leave it to the user to do the normalization.

Reply:
Alright; will make sure the docstring calls this out as explicitly as possible.